File: src/creational/abstract-factory/python/classes/factories/customer_vehicle_factory.py
from classes.customer import Customer
from classes.customer_vehicle import CustomerVehicle
from classes.interfaces.user_protocol import UserProtocol
from classes.interfaces.user_vehicle_factory_protocol import \
    UserVehicleFactoryProtocol
from classes.interfaces.vehicle_protocol import VehicleProtocol


class CustomerVehicleFactory(UserVehicleFactoryProtocol):
    @staticmethod
    def create_user(name: str, location: str) -> UserProtocol:
        return Customer(name, location)

    @staticmethod
    def create_vehicle(name: str) -> VehicleProtocol:
        return CustomerVehicle(name)

File: src/creational/abstract-factory/typescript/classes/factories/CustomerVehicleFactory.ts
import { Customer } from '../Customer';
import { CustomerVehicle } from '../CustomerVehicle';
import { UserProtocol } from '../interfaces/UserProtocol';
import { UserVehicleFactoryProtocol } from '../interfaces/UserVehicleFactoryProtocol';
import { VehicleProtocol } from '../interfaces/VehicleProtocol';

export class CustomerVehicleFactory implements UserVehicleFactoryProtocol {
  createUser(name: string, location: string): UserProtocol {
    return new Customer(name, location);
  }
  createVehicle(name: string): VehicleProtocol {
    return new CustomerVehicle(name);
  }
}
